src/compiler/scala/tools/nsc/transform/CleanUp.scala

```diff
@@ -383,6 +380,106 @@ abstract class CleanUp extends Statics with Transform with ast.TreeDSL {
+      sw.selector.tpe match {
+        case IntTpe => sw // can switch directly on ints
+        case StringTpe =>
+          // these assumptions about the shape of the tree are justified by the codegen in MatchOptimization
+          val Match(Typed(selTree: Ident, _), cases) = sw
+          val sel = selTree.symbol
+          val restpe = sw.tpe
+          val swPos = sw.pos.focus
+
+          /* From this:
+           *     string match { case "AaAa" => 1 case "BBBB" | "c" => 2 case _ => 3}
+           * Generate this:
+           *     goto matchSuccess (string.## match {
+           *       case 2031744 =>
+           *         if ("AaAa" equals string) goto match1
+           *         else if ("BBBB" equals string) goto match2
+           *         else goto matchFailure
+           *       case 99 =>
+           *         if ("c" equals string) goto match2
+           *         else goto matchFailure
+           *       case _ => goto matchFailure
+           *     }
+           *     match1: goto matchSuccess (1)
+           *     match2: goto matchSuccess (2)
+           *     matchFailure: goto matchSuccess (3) // would be throw new MatchError(string) if no default was given
+           *     matchSuccess(res: Int): res
+           * This proliferation of labels is needed to handle alternative patterns, since multiple branches in the
+           * resulting switch may need to correspond to a single case body.
+           */
+
+          val stats = mutable.ListBuffer.empty[Tree]
+          var failureBody = Throw(New(definitions.MatchErrorClass.tpe_*, REF(sel))) : Tree
+
+          // genbcode isn't thrilled about seeing labels with Unit arguments, so `success`'s type is one of
+          // `${sw.tpe} => ${sw.tpe}` or `() => Unit` depending.
+          val success = {
+            val lab = currentOwner.newLabel(unit.freshTermName("matchEnd"), swPos)
+            if (restpe =:= UnitTpe) {
+              lab.setInfo(MethodType(Nil, restpe))
+            } else {
+              lab.setInfo(MethodType(lab.newValueParameter(nme.x_1).setInfo(restpe) :: Nil, restpe))
+            }
+          }
+          def succeed(res: Tree): Tree =
+            if (restpe =:= UnitTpe) BLOCK(res, REF(success) APPLY Nil) else REF(success) APPLY res
+
+          val failure = currentOwner.newLabel(unit.freshTermName("matchEnd"), swPos).setInfo(MethodType(Nil, restpe))
+          def fail(): Tree = atPos(swPos) { Apply(REF(failure), Nil) }
+
+          val newSel = atPos(sel.pos) { IF (sel OBJ_EQ NULL) THEN LIT(0) ELSE (Apply(REF(sel) DOT Object_hashCode, Nil)) }
+          val casesByHash =
+            cases.flatMap {
+              case cd@CaseDef(StringsPattern(strs), _, body) =>
+                val jump = currentOwner.newLabel(unit.freshTermName("case"), swPos).setInfo(MethodType(Nil, restpe))
+                stats += LabelDef(jump, Nil, succeed(body))
+                strs.map((_, jump, cd.pat.pos))
```

Comment:
I used to copy the body of a case if there were alternatives, so
```scala
x match { case "a" | "b" => body }
```
would introduce tree sharing.